### PR TITLE
Global performance improvements

### PR DESCRIPTION
**Pull request by pavel-kirienko**
DO NOT MERGE

This is still WIP. Changes:

- Debug/Nuttx.py:
  - Now it works on GDB 7.6 with Python 2.7 ([from Terry Guo et al](https://launchpad.net/gcc-arm-embedded)).
  - Output format of `show mybt` changed to match GDB's bt default format.
  - `show mybt` still can't unwind the call stack correctly though - it was broken and it remains this way.
- Added a simple sampling profiler Debug/poor-mans-profiler.sh.
- Stack overflow checks via `__cyg_profile_func*()` are disabled by default, per @LorenzMeier's suggestion. They can be enabled back by setting `ENABLE_STACK_CHECKS=1`, e.g. `make ENABLE_STACK_CHECKS=1`.
- Enabled extended debug information generation: `-g` replaced with `-g3`. This is not necessary, but it helped me with profiling and it doesn't affect the final binary, so it should be safe to keep.
- Changes to the uavcan module (see https://github.com/PX4/Firmware/issues/1417 for details):
  - More perfcounters
  - Stack overflow checks are always disabled, regardless of `ENABLE_STACK_CHECKS`
  - `-O3` instead of `-Os`

One thing left to fix: the current head of this branch crashes occasionally (50/50) shortly *after* boot:

```
uavcan: subscribe to actuator_controls_0
navigator: Initializing..
mavlink_if0: DISABLING HARDWARE FLOW CONTROL
[i] Addons script: /fs/microsd/etc/extras.txt 

NuttShell (NSH)
nsh> navigator: Could not clear geofence
navigator: offboard mission updated: dataman_id=0, count=0, current_seq=0
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: uavcan
sp:     20009264
IRQ stack:
  base: 200092d0
  size: 00000ffc
20009260: 200139ac 20009264 20013af0 200092d0 08078931 000000b8 080a9fd8 200139ac
20009280: 08078b61 08086645 08078a01 00000010 08078c05 08078be9 08089751 200065e4
200092a0: 00000003 200139ac 08086671 00000000 20015930 20011a10 00000008 20011a10
200092c0: 20011938 00000001 08078b61 200139ac 00000000 10000010 00000001 00000000
sp:     20013a80
User stack:
  base: 20013af0
  size: 00001384
20013a80: 00000000 00000000 00000000 00000000 08059b6d 20015930 00000000 08059a17
20013aa0: 080599f5 2000da60 00000004 20011570 20011a08 0805b433 2000da60 0001209c
20013ac0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 0805b58b
20013ae0: 0805b581 0807150d 00000000 00000000 ffffffff 198aed12 00000040 80001390
R0: 00000000 20011a10 00000000 63696d2f 20015930 20011a10 00000008 20011a10
R8: 20011938 00000001 200127b0 20011874 00000000 20013a80 08059861 08059c5e
xPSR: 61000000 BASEPRI: 00000000 CONTROL: 00000000
EXC_RETURN: ffffffe9
```

LR: modules/systemlib/mixer/mixer.cpp:73  
PC: modules/systemlib/mixer/mixer_multirotor.cpp:217

Bisecting through the history, I found that the breaking commits are aa36686e765b11da040a536326aa8977ffd87d81 and 530988dfd1997f09c2deb61890859aa8d4399114 (the former introduced another bug that was fixed in the latter), so the change that broke the module was adding `-fno-instrument-functions` to the uavcan module.

I'm currently half asleep and am likely missing something obvious here - I'll continue tomorrow. Meanwhile, any pointers or comments would be appreciated.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
I don't think you can partially disable instrumentation. Its all or nothing. So I would suggest to remove the per-module instrumentation and only allow the global instrumentation. I can't see a way to do it per module, because the stack checker would need to know in which module it operates, which it doesn't.

**Comment by pavel-kirienko** [comment]
> I can't see a way to do it per module, because the stack checker would need to know in which module it operates, which it doesn't.

Could you elaborate on this? Why does the stack checker need to know the module it is invoked from? According to the source, it is very basic and unaware of the calling module: https://github.com/PX4/NuttX/blob/5ee4b2b2c26bbc35d1669840f0676e8aa383b984/nuttx/arch/arm/src/armv7-m/up_stackcheck.c#L31

Moreover, it fails even if the whole application is compiled without `-finstrument-functions`. As I observed in https://github.com/PX4/Firmware/issues/1417, something is still calling the stack checker even when the application has it disabled, probably it's NuttX itself. Do I need to disable it in NuttX as well?

**Comment by LorenzMeier** [comment]
Because the stack checker reads the content or R10. Which means that R10 does need to be reserved by *each and every line of code* and that R10 needs to be initialized correctly *by each and every module*. So its really everything or nothing, and yes, this also applies to the OS.

**Comment by pavel-kirienko** [comment]
As I understand, r10 is used to keep stack limit for the current task, and it is initialized regardless of `-finstrument-functions`. Is it not correct?

I'm asking because all modules are compiled with `-ffixed-r10` regardless of the stack check feature, so I don't see how it can fail unless my understanding above is incorrect.

**Comment by pavel-kirienko** [comment]
Aha, seems like my understanding was a bit off after all. I recompiled the OS with stack checks disabled (and r10 unfixed), and it seems to work fine so far (5 reboots in a row without failures). CPU usage has dropped dramatically:

```
Processes: 21 total, 2 running, 19 sleeping
CPU usage: 36.08% tasks, 0.56% sched, 63.36% idle
Uptime: 20.086s total, 11.976s idle

 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) STATE 
   0 Idle Task                   11975 63.362     0/    0   0 (  0)  READY 
   1 hpwork                        174  0.951   580/ 1992 192 (192)  w:sig 
   2 lpwork                         69  0.317   572/ 1992  50 ( 50)  READY 
   3 init                         1892  0.000  1452/ 2992 100 (100)  w:sem 
   6 nshterm                         0  0.000   500/ 1592  70 ( 70)  w:sem 
 147 top                           138  2.061  1148/ 1696 100 (100)  RUN   
  59 dataman                        33  0.000   652/ 1992 100 (100)  w:sem 
  61 commander                      96  0.158  2588/ 3192 215 (215)  w:sig 
  62 commander_low_prio              2  0.000   468/ 2896  50 ( 50)  w:sem 
  69 px4io                         489  2.775   868/ 1992 240 (240)  w:sem 
  79 mavlink_if0                   131  0.713  1748/ 2792 100 (100)  w:sig 
  80 mavlink_rcv_if0                 1  0.000   764/ 2896 175 (175)  w:sem 
  85 uavcan                        520  3.092  1020/ 2992 240 (240)  w:sem 
 116 sensors_task                  405  2.299  1068/ 1992 250 (250)  w:sem 
 121 sdlog2                         70  0.317  1884/ 2992  70 ( 70)  w:sig 
 124 gps                            32  0.000   700/ 1496 220 (220)  w:sem 
 136 attitude_estimator_ekf       3032 18.160  6252/ 7696 250 (250)  w:sem 
 138 position_estimator_inav       259  1.665  4140/ 4992 250 (250)  w:sem 
 140 mc_att_control                426  2.537  1052/ 1992 250 (250)  w:sem 
 142 mc_pos_control                 60  0.396   868/ 1992 250 (250)  w:sem 
 146 navigator                      99  0.634   796/ 1792 120 (120)  w:sem 
```

uavcan - 3% vs. 14%

total - 36% vs. 76%

**Comment by LorenzMeier** [comment]
I have to admit I had recalled a number of about 20% overhead. Its quite possible that we've restructured things a bit since then and have now more function calls on average. Looks like we should get this in as soon as possible - do you have stack checking enabling / disabling now global for NuttX and app?

Your reasoning that correct init of R10 and partial checking should work makes sense, however, I presume that something in the compile logic failed not fixing R10 or allocating stack without initializing R10.

File: src/modules/uavcan/actuators/esc.cpp [hunk 1]
```diff
@@ -49,12 +49,24 @@ UavcanEscController::UavcanEscController(uavcan::INode &node) :
...
+	if (_perfcnt_broadcast_elapsed == nullptr) {
+		errx(1, "uavcan: couldn't allocate _perfcnt_broadcast_elapsed");
+	}
 }
 
 UavcanEscController::~UavcanEscController()
 {
 	perf_free(_perfcnt_invalid_input);
 	perf_free(_perfcnt_scaling_error);
+	perf_free(_perfcnt_broadcast_elapsed);
```
Comment (LorenzMeier) [inline src/modules/uavcan/actuators/esc.cpp]: We start to have an enormous amount of perf counters, see https://github.com/PX4/Firmware/issues/1623. I'm not against adding the ones we need, but please consolidate them as far as sensible. Also keep in mind that the strings used there use space, so shorter names (as long as human readable) are preferred.

**Comment by pavel-kirienko** [comment]
It failed just now again. It's a floating bug so it's hard to reliably determine when it's gone.

I tried to disable stack checker globally ([here](https://github.com/PX4/Firmware/blob/master/nuttx-configs/px4fmu-v2/nsh/Make.defs#L65)), both with R10 fixed and relaxed, and it failed in both configurations if I tried hard enough.

File: src/modules/uavcan/actuators/esc.cpp (hunk 1 above)
Comment (pavel-kirienko) [inline src/modules/uavcan/actuators/esc.cpp]: Sure, most of them can be removed. Why their number is a concern? According to my profiling results, their performance impact is negligible.

**Comment by LorenzMeier** [comment]
The stack trace doesn't hint at a specific memory region / location, right? It could be well also a real bug in need of fixing.

**Comment by pavel-kirienko** [comment]
> The stack trace doesn't hint at a specific memory region / location, right?

No.

$ arm-none-eabi-addr2line -e Build/px4fmu-v2_default.build/firmware.elf 08075484
/media/storage/px4/Firmware/NuttX/nuttx/mm/mm_shrinkchunk.c:80
$ arm-none-eabi-addr2line -e Build/px4fmu-v2_default.build/firmware.elf 08075265
/media/storage/px4/Firmware/NuttX/nuttx/mm/mm_realloc.c:147

File: src/modules/uavcan/actuators/esc.cpp (hunk 1 above)
Comment (LorenzMeier) [inline src/modules/uavcan/actuators/esc.cpp]: On a full system you can't fit them on a screen and their strings eat flash space. I'm trying to also remove the boiler-plate shell handing and error handling code. That should allow us to recuperate a reasonable amount of space, allowing us to live well in our 1M prison (thanks ST!).

**Comment by pavel-kirienko** [comment]
I have to attend some IRL business now and will continue later tonight and tomorrow. Meanwhile, it would be appreciated if somebody else could give more pointers or ask the right questions.

BTW, @LorenzMeier, image size with stack checker globally disabled, R10 fixed and uavcan with -O3 is ~748K.

**Comment by pavel-kirienko** [comment]
I did a number of tests on the off chance that the issue can be caused by the optimization. I managed to make it fail at least once on all optimization levels (1, 2, 3, s). Also, it seems like it is easier to trigger a failure via power on reset, rather than via `reboot`.

This is where it fails:
- Firmware/src/modules/systemlib/mixer/mixer_group.cpp:107 (task "uavcan")
- Firmware/NuttX/nuttx/mm/mm_shrinkchunk.c:80 (task "init" or "uavcan", called from realloc())

This looks like heap damage so far.

**Comment by LorenzMeier** [comment]
Does the issue persist if you comment out 'mixer load' in the rc script? Should be in rc.interface. Wondering if this is a string handling issue that you've uncovered.

**Comment by pavel-kirienko** [comment]
Will check this a bit later. Currently I'm trying to isolate the exact change that introduced the failure.

**Comment by pavel-kirienko** [comment]
It's getting weirder. In order to confirm that the breaking change is the removal of stack checks and nothing else, I switched to master (about 2 days old) and removed the checks:

```diff
diff --git a/makefiles/toolchain_gnu-arm-eabi.mk b/makefiles/toolchain_gnu-arm-eabi.mk
index d7e48f0..3cc2626 100644
--- a/makefiles/toolchain_gnu-arm-eabi.mk
+++ b/makefiles/toolchain_gnu-arm-eabi.mk
@@ -80,11 +80,9 @@ ARCHCPUFLAGS_CORTEXM3         = -mcpu=cortex-m3 \
                           -march=armv7-m \
                           -mfloat-abi=soft
 
-ARCHINSTRUMENTATIONDEFINES_CORTEXM4F = -finstrument-functions \
-                          -ffixed-r10
+ARCHINSTRUMENTATIONDEFINES_CORTEXM4F = 
 
-ARCHINSTRUMENTATIONDEFINES_CORTEXM4 = -finstrument-functions \
-                          -ffixed-r10
+ARCHINSTRUMENTATIONDEFINES_CORTEXM4 =
 
 ARCHINSTRUMENTATIONDEFINES_CORTEXM3 =
 
diff --git a/nuttx-configs/px4fmu-v2/nsh/Make.defs b/nuttx-configs/px4fmu-v2/nsh/Make.defs
index 99f3b31..3ab8a16 100644
--- a/nuttx-configs/px4fmu-v2/nsh/Make.defs
+++ b/nuttx-configs/px4fmu-v2/nsh/Make.defs
@@ -62,8 +62,7 @@ ARCHCPUFLAGS           = -mcpu=cortex-m4 \
 
 
 # enable precise stack overflow tracking
-INSTRUMENTATIONDEFINES  = -finstrument-functions \
-                          -ffixed-r10
+INSTRUMENTATIONDEFINES  = 
 
 # pull in *just* libm from the toolchain ... this is grody
 LIBM                    = "${shell $(CC) $(ARCHCPUFLAGS) -print-file-name=libm.a}"
```

This compiled successfully, I flashed the board, started it and noticed that it can't communicate with PX4IO. This is what happened:

```
check CRC failed - -1
px4io: px4io is not started, still attempting upgrade
[PX4IO] using firmware from /etc/extras/px4io-v2_default.bin
[PX4IO] found bootloader revision: 4
[PX4IO] erase...
[PX4IO] programming 42696 bytes...
[PX4IO] verify...
[PX4IO] update complete
check CRC failed - -1
[i] ERROR: PX4IO not found 
<...>
[px4io] failed to communicate with IO
px4io: driver init failed
[i] ERROR: PX4IO start failed
```

The issue can be reliably reproduced on every reboot. Re-enabling stack checks makes it communicate with PX4IO fine again.

**Comment by pavel-kirienko** [comment]
> Does the issue persist if you comment out 'mixer load' in the rc script?

@LorenzMeier No, the issue does not persist if the mixer is not loaded. Since the failure can be triggered both from `uavcan` and `init`, and since the mixer appears to be the only shared module between these tasks, I suspect that the problem is probably somewhere inside the mixer module. Though I'm not sure how the PX4IO update issue described above relates to this... but then again, if it's a heap corruption, it can affect anything.

**Comment by davids5** [comment]
This is the same thing I was seeing on the nuttx_bringup branch. On the current Nuttx commit. I reverted to NuttX 007dd93e6c3b99319ae0925846e7b44d50ac5275. A the px4io update worked fine. Then after returning to the current Nuttx commit. The problem was gone. 

So something was squirrely with the loading of the px4io firmware from the FMU. When I went to debug what I thought would be a crashed px4io, but instead I found that the stack pointer, reset vector had 0xffffff. <ip>

I downloaded masters boot loader to the px4io. Ran the test the second time It it failed to identify the px4io. 

I set breakpoints in the fmu code and checked all the return status and CRC was fine. 

The the problem just vanished.

We will need to dig deeper.

**Comment by LorenzMeier** [comment]
Ok Pavel. The mixer loading has some string operations going on, I guess we'll need to dig into that a bit deeper then.

**Comment by pavel-kirienko** [comment]
I need to switch context now, so I'm putting this issue on freeze. Also, it's blocked by https://github.com/PX4/Firmware/issues/1675.

**Comment by davids5** [comment]
@LorenzMeier @pavel-kirienko  

No more hard faults! The ppx4io may not program on the first boot or subsequent command line invocation. There is still some timing issues there. Sometimes it works. I will need to look deeper into this.

I used the boards exported config.h to set the ENABLE_STACK_CHECKS in toolchain_gnu-arm-eabi.mk. I used grep on the config.h instead of just including the defconfig. The reasoning is: 1) A normal nuttx build would use make oldconfig, make menuconfig. These 2 operations may change the state of a setting in the incoming defconfig.
2) If the nuttx exports were not build locally (just pulled from a server) the only true way to know state of CONFIG_ARMV7M_STACKCHECK for that build is in config.h

If having grep in the tools is an issues, we could also export the ,config file to the nuttx-export of a given board and then source that file into the board_xxxx.mk file. The only issue I see with that is there could be a collision.

**Comment by LorenzMeier** [comment]
Thanks! I think I want to re-organize the startup logic a bit to make it less timing sensitive. What is happening is that we're trying to talk to the PX4IO application code before its ready, and so all it takes is to either re-arrange the startup sequencing, and/or to allow the code to either talk to application level or bootloader code.

**Comment by LorenzMeier** [comment]
@davids5 @pavel-kirienko Will do one last test round here, then merge.

**Comment by pavel-kirienko** [comment]
Great, I don't think I have anything to add here.

**Comment by LorenzMeier** [comment]
Test came clean, merged!

**Pull request merged**